### PR TITLE
githuboauth: Check all existing emails before creating account

### PR DESCRIPTION
**Pull request by ryanslade**
If a we allow user sign ups we should check all existing email addresses
before deciding to create a new account.

The previous code would create a new account only if the primary address
didn't match and wouldn't fall back the secondary addresses.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/auth/githuboauth/session.go [hunk 1]
```diff
@@ -72,11 +72,36 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
...
 	)
-	for i, verifiedEmail := range verifiedEmails {
+
+	// We will first attempt to connect one of the verified emails with an existing
+	// account in Sourcegraph
+	type attemptConfig struct {
+		email            string
+		createIfNotExist bool
+	}
+	var attempts []attemptConfig
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/githuboauth/session.go:82]: Nice 👍

File: enterprise/cmd/frontend/internal/auth/githuboauth/session_test.go [hunk 2]
```diff
@@ -247,6 +247,61 @@ func TestSessionIssuerHelper_GetOrCreateUser(t *testing.T) {
...
+	}
+
+	ctx := githublogin.WithUser(context.Background(), ghUser)
+	s := &sessionIssuerHelper{
+		CodeHost:    codeHost,
+		clientID:    clientID,
+		allowSignup: true,
+		allowOrgs:   nil,
+	}
+	tok := &oauth2.Token{AccessToken: "dummy-value-that-isnt-relevant-to-unit-correctness"}
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/githuboauth/session_test.go:298]: 😂

**Review by unknwon — APPROVED** [review]
Assuming this has been tested e2e.

**Comment by ryanslade** [comment]
I confirmed this works by doing the following:

1. Created an account manually using my *secondary* email in GitHub
2. Logged out
3. Signup up with github.com and confirmed that I was connected with my existing account

**Pull request merged**